keeper/upkeep_executer.go

```diff
@@ -0,0 +1,280 @@
+	"github.com/smartcontractkit/external-initiator/keeper/keeper_registry_contract"
+	"github.com/smartcontractkit/external-initiator/store"
+	"go.uber.org/atomic"
+)
+
+const (
+	checkUpkeep        = "checkUpkeep"
+	performUpkeep      = "performUpkeep"
+	executionQueueSize = 10
+)
+
+var (
+	upkeepRegistryABI = mustGetABI(keeper_registry_contract.KeeperRegistryContractABI)
+	performUpkeepHex  = utils.AddHexPrefix(common.Bytes2Hex(upkeepRegistryABI.Methods[performUpkeep].ID))
+	refreshInterval   = 5 * time.Second
+)
+
+type UpkeepExecuter interface {
+	Start() error
+	Stop()
+}
+
+func NewUpkeepExecuter(dbClient *gorm.DB, clNode chainlink.Node, config store.RuntimeConfig) UpkeepExecuter {
+	return upkeepExecuter{
+		blockHeight:    atomic.NewUint64(0),
+		chainlinkNode:  clNode,
+		endpoint:       config.KeeperEthEndpoint,
+		registryStore:  NewRegistryStore(dbClient, uint64(config.KeeperBlockCooldown)),
+		executionQueue: make(chan struct{}, executionQueueSize),
+		chDone:         make(chan struct{}),
+		chSignalRun:    make(chan struct{}, 1),
+	}
+}
+
+type upkeepExecuter struct {
+	blockHeight   *atomic.Uint64
+	chainlinkNode chainlink.Node
+	endpoint      string
+	ethClient     *ethclient.Client
+	registryStore RegistryStore
+
+	executionQueue chan struct{}
+	chDone         chan struct{}
+	chSignalRun    chan struct{}
+}
+
+var _ UpkeepExecuter = upkeepExecuter{} // upkeepExecuter satisfies UpkeepExecuter
+
+func (executer upkeepExecuter) Start() error {
+	ethClient, err := ethclient.Dial(executer.endpoint)
+	if err != nil {
+		return err
+	}
+	executer.ethClient = ethClient
+
+	if strings.HasPrefix(executer.endpoint, "ws") {
+		go executer.setRunsOnHeadSubscription()
+	} else if strings.HasPrefix(executer.endpoint, "http") {
+		go executer.setRunsOnTimer()
+	} else {
+		return fmt.Errorf("unknown endpoint protocol: %+v", executer.endpoint)
+	}
+
+	go executer.run()
+
+	return nil
+}
+
+func (executer upkeepExecuter) Stop() {
+	close(executer.chDone)
+}
+
+func (executer upkeepExecuter) run() {
+	for {
+		select {
+		case <-executer.chDone:
+			return
+		case <-executer.chSignalRun:
+			executer.processActiveRegistrations()
+		}
+	}
+}
+
+func (executer upkeepExecuter) processActiveRegistrations() {
+	// TODO - RYAN - this should be batched to avoid congestgion
+	logger.Debug("received new block, running checkUpkeep for keeper registrations")
+
+	activeRegistrations, err := executer.registryStore.Active(executer.blockHeight.Load())
+	if err != nil {
+		logger.Errorf("unable to load active registrations: %v", err)
+		return
+	}
+
+	for _, reg := range activeRegistrations {
+		executer.concurrentExecute(reg)
+	}
+}
+
+func (executer upkeepExecuter) concurrentExecute(registration registration) {
+	executer.executionQueue <- struct{}{}
+	go executer.execute(registration)
+}
+
+// execute will call checkForUpkeep and, if it succeeds, triger a job on the CL node
+func (executer upkeepExecuter) execute(registration registration) {
```

Comment:
Lots going on in this method. I think it might help to break out a few smaller helper methods to organize and document... possibly easier to test too.